# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [79]:
import pandas as pd

In [80]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing" 
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" 
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)
orderlines

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**. 

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method. 

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [81]:
# orders
orders.duplicated().sum()

0

In [82]:
# orderlines
orderlines.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

# 2.&nbsp; `.info()`

In [83]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [84]:
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [85]:
print(f"5 missing values represents {((orders.total_paid.isna().sum() / orders.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [86]:
orders = orders.loc[~orders.total_paid.isna(), :]
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226904 non-null  int64  
 1   created_date  226904 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226904 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 8.7+ MB


Should you have a significant number of missing values in the future, you have a choice: 
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [87]:
orders["created_date"] = pd.to_datetime(orders["created_date"])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [88]:
orderlines["date"] = pd.to_datetime(orderlines["date"])

#### 4.1.2.&nbsp;`unit_price`

In [89]:
#orderlines["unit_price"] = pd.to_numeric(orderlines["unit_price"])

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [90]:
orderlines.unit_price.astype(str).str.contains("\d+\.\d+\.\d+").value_counts()

False    257814
True      36169
Name: unit_price, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [91]:
two_dot_percentage = ((orderlines.unit_price.astype(str).str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [92]:
two_dot_order_ids_list = orderlines.loc[orderlines.unit_price.astype(str).str.contains("\d+\.\d+\.\d+"), "id_order"]
orderlines = orderlines.loc[~orderlines.id_order.isin(two_dot_order_ids_list)]

In [93]:
orderlines.shape[0]

216250

We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [94]:
orderlines["unit_price"] = pd.to_numeric(orderlines["unit_price"])


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


It worked perfectly

# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [95]:
# products.csv
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing" 
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv(path)


In [96]:
products.info() # Total No of rows and columns are (19326 rows, 7 Columns)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


### Look for Duplicates

In [97]:
products.duplicated().sum() #Counts of duplicated rows in products table 
#Duplicated rows are 8746

8746

In [98]:
products[products.duplicated()] # list of duplicated dta entries in peolpe table


sku                                               name  \
101    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
102    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
103    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
104    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
105    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
...        ...                                                ...   
16831  APP2302  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16833  APP2303  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
18190  PAR0077                         Parrot Bebop Drone 2 Power   
18308  NKI0010              Nokia Wireless sphygmomanometer Plata   
18514  PLA0035          Bluetooth Headset Plantronics Explorer 80   

                                                    desc        price  \
101          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
102          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
103          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
104          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
105          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
...                                                  ...          ...   
16831  New MacBook Pro 13-inch Core i5 Touch Bar 33 G...   26.155.941   
16833  New MacBook Pro 13 inch Touch Bar 33 GHz Core ...  237.559.421   
18190  Drone cuadricóptero quality camera integrated ...        699.9   
18308     Sphygmomanometer for iPhone iPad and iPod App.       129.99   
18514  egonómico wireless headset design for iPhone i...        39.99   

      promo_price  in_stock      type  
101    11.455.917         0      1282  
102    11.455.917         0      1282  
103    11.455.917         0      1282  
104    11.455.917         0      1282  
105    11.455.917         0      1282  
...           ...       ...       ...  
16831  26.155.941         0  1,02E+12  
16833  23.755.942         0  1,02E+12  
18190   6.733.892         0  11905404  
18308   1.149.899         1  11905404  
18514     349.896         1      5384  

[8746 rows x 7 columns]

In [99]:
products=products.drop_duplicates(subset=['sku']) # Removal of duplicated rows comparing with 'sku' column while it is primary key for products table
products # output after removing duplicated data is (10579 rows × 7 columns)

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19321         1     12282  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[10579 rows x 7 columns]

### Look for Missing values


In [100]:
products.price.isna().sum() #No. of missing value row count for price column


45

In [101]:
products.type.isna().sum()

50

In [102]:
products.loc[products.price.isna(), :] # List of above missing values

sku                                               name  \
34       TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900     AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039     CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042     CEL0007  Celly Wallet Case with removable cover Black i...   
2043     CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   
2044     CEL0014        Celly Silicone Hard Shell iPhone 6 Amarillo   
2049     CEL0015    Celly fur-lined Powerbank battery 4000mAh Black   
2051     CEL0018     Celly Wallet Leather Case cover Black iPhone 6   
2052     CEL0023  Celly Ambo Luxury Leather Case + Case Gold iPh...   
2053     CEL0025  Celly Ambo Luxury Leather Case + Case iPhone 6...   
12052    CEL0028  Celly Swing Arm button selfie with Bluetooth B...   
12053    CEL0029  Celly extensible arm selfie with Bluetooth but...   
12054    CEL0030  Celly Smart Drive Housing + clip car holder iP...   
12055    CEL0031  Celly Smart Drive Case + Car clip support iPho...   
12059    CEL0032      Celly Gost clamp fixing car + Magnetica Black   
12060    CEL0042  Celly iPhone Case Fluor 6s / 6 Black / Transpa...   
12061    CEL0043  Celly Fluor Case iPhone 6 / 6S Pink / Transparent   
12062    CEL0044  Celly iPhone Case Fluor 6s / 6 Yellow / Transp...   
12064    CEL0038           Celly Frost Cover for iPhone 6 / 6S Rosa   
12067    CEL0033  Celly Waterproof iPad Case / Air / Pro 97 "Tra...   
12068    CEL0034  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12069    CEL0035  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12070    CEL0036  Celly Adapter Micro USB connector to Lightning...   
12071    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
12072    CEL0052   Celly 4000mAh Battery Power Bank Aluminum Silver   
12073    CEL0053             Celly IceCube Case iPhone 6 / 6S Black   
12074    CEL0054              Celly IceCube Case iPhone 6 / 6S Rosa   
12075    CEL0055             Celly IceCube Case iPhone 6 / 6S White   
12076    CEL0056          IceCube Celly iPhone Case 6 / 6s Amarillo   
12077    CEL0057        Celly IceCube Case iPhone 6 / 6S Plus Black   
12078    CEL0058         Celly IceCube Case iPhone 6 / 6S Plus Rosa   
12079    CEL0059        Celly IceCube Case iPhone 6 / 6S Plus White   
12080    CEL0060     Celly IceCube Case iPhone 6 / 6S Plus Amarillo   
12088    CEL0040     Celly Frost Cover for iPhone 6 / 6S Plus White   
12089    CEL0041              Celly iPhone 6S Plus Frost Case Black   
12092    CEL0045         Laser Case Celly iPhone 6 / 6S Dark Silver   
12093    CEL0046        Laser Case Celly iPhone 6 / 6S Clear Silver   
12095    CEL0047              Laser Case Celly iPhone 6 / 6S Dorado   
12096    CEL0048                Laser Case Celly iPhone 6 / 6S Rosa   
12097    CEL0049         Laser Case Celly iPhone 6 / 6S Plus Dorado   
12098    CEL0050   Laser Case Celly iPhone 6 / 6S Plus Silver Clear   
12100    CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12238    CEL0027          Celly Frost Cover for iPhone 6 / 6S White   
12242    CEL0039       Celly Frost Cover for iPhone 6 / 6S Amarillo   
12947  CEL0034-A    (Open) Celly iPhone 6S Plus Housing Frost White   

                                                    desc price promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...   NaN     299.899   
1900       MacBook Air 11-inch casing with matte finish.   NaN       22.99   
2039   Cover and housing together with magnet for iPh...   NaN     399.905   
2042              Case Book for iPhone 6 card case type.   NaN     128.998   
2043                       Hard Shell Silicone iPhone 6.   NaN        4.99   
2044                       Hard Shell Silicone iPhone 6.   NaN      59.895   
2049   Leather-wrapped External Battery 4000mAh for i...   NaN     239.895   
2051   Card case with transparent protective cover fo...   NaN     294.877   
2052   Cover and housing together with magne

In [103]:
products = products.loc[~products.price.isna(), :] #removing rows from products table for which price is null
#products = products.loc[~products.desc.isna(), :] # Will check later is any business question is related with this field
products = products.loc[~products.type.isna(), :] # Will check later is any business question is related with this field
products.info() # Output after removing missing price values and type and drop promo in products table (Rows-10110  column 6)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10484 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10484 non-null  object
 1   name         10484 non-null  object
 2   desc         10477 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10484 non-null  object
 5   in_stock     10484 non-null  int64 
 6   type         10484 non-null  object
dtypes: int64(1), object(6)
memory usage: 655.2+ KB


### Check / Change Data types

In [104]:
#products["price"] = pd.to_numeric(products["price"]) # price column contains 1.639.792 value

In [105]:
products.price.astype(str).str.contains("\d+\.\d+\.\d+").value_counts() # Calculate how many columns are affected with two dots values -377 

False    10110
True       374
Name: price, dtype: int64

In [106]:
two_dot_percentage = ((products.price.astype(str).str.contains("\d+\.\d+\.\d+").value_counts()[1] / products.shape[0])*100).round(2) # total percentage of overall data-3.58%
two_dot_percentage

3.57

In [107]:
two_dot_sku_list =products.loc[products.price.astype(str).str.contains("\d+\.\d+\.\d+"), "sku"]
products = products.loc[~products.sku.isin(two_dot_sku_list)] # removal of dwo dotted price values from products table
products.shape[0] #Total no. of available rowas after removal are 10110


10110

In [108]:
products["price"] = pd.to_numeric(products["price"]) # Change the datatype of rest price values from object to float
products.info() # price column Data type has been changed to float


<class 'pandas.core.frame.DataFrame'>
Int64Index: 10110 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10110 non-null  object 
 1   name         10110 non-null  object 
 2   desc         10104 non-null  object 
 3   price        10110 non-null  float64
 4   promo_price  10110 non-null  object 
 5   in_stock     10110 non-null  int64  
 6   type         10110 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 631.9+ KB


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [109]:
promo_price_number = products.loc[(products.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
promo_price_number # No. of rows affected by weird pattern in promo price column

9346

In [110]:
print(f"The column promo_price has in total {promo_price_number} wrong values. This is {round(((promo_price_number / products.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column promo_price has in total 9346 wrong values. This is 92.44% of the rows of the DataFrame


In [111]:
products = products.drop(columns=["promo_price"]) # solution may be we can just drop the promo_price from our dataframe
products # Output after cleaning the data we have 10110 rows × 6 columns

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price  in_stock  \
0           Aluminum support compatible with all MacBook  59.99         1   
1              USB ultrathin keyboard Apple Mac Spanish.  59.00         0   
2                                 mouse Apple USB cable.  59.00         0   
3                  IPhone dock and USB Cable Apple iPod.  25.00         0   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99         1   
...                                                  ...    ...       ...   
19321  compact and portable stand vertically or horiz...  29.99         1   
19322  Backpack with capacity of 14 liter compartment...  69.95         1   
19323  Backpack with capacity of 14 liter compartment...  69.95         1   
19324  Backpack with capacity of 14 liter compartment...  69.95         0   
19325  Backpack with capacity of 14 liter compartment...  69.95         1   

           type  
0          8696  
1      13855401  
2          1387  
3          1230  
4          1364  
...         ...  
19321     12282  
19322      1392  
19323      1392  
19324      1392  
19325      1392  

[10110 rows x 6 columns]

In [112]:
from google.colab import files #cleaned version of products table

products.to_csv("products.csv", index=False)
files.download("products.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [113]:
#ordline_products = pd.merge(orderlines,products, how = "outer", on = ['sku'])
#ordline_products

In [114]:
#ordline_products.isna().sum()

In [115]:
merge_orderline_products = products.merge(orderlines)
merge_orderline_products

sku                                      name  \
0       RAI0007         Silver Rain Design mStand Support   
1       RAI0007         Silver Rain Design mStand Support   
2       RAI0007         Silver Rain Design mStand Support   
3       RAI0007         Silver Rain Design mStand Support   
4       RAI0007         Silver Rain Design mStand Support   
...         ...                                       ...   
212551  REP0076  repair Full screen iPad (1st generation)   
212552  REP0076  repair Full screen iPad (1st generation)   
212553  REP0076  repair Full screen iPad (1st generation)   
212554  REP0076  repair Full screen iPad (1st generation)   
212555  REP0076  repair Full screen iPad (1st generation)   

                                                     desc   price  in_stock  \
0            Aluminum support compatible with all MacBook   59.99         1   
1            Aluminum support compatible with all MacBook   59.99         1   
2            Aluminum support compatible with all MacBook   59.99         1   
3            Aluminum support compatible with all MacBook   59.99         1   
4            Aluminum support compatible with all MacBook   59.99         1   
...                                                   ...     ...       ...   
212551  Repair service including parts and labor for iPad  149.99         0   
212552  Repair service including parts and labor for iPad  149.99         0   
212553  Repair service including parts and labor for iPad  149.99         0   
212554  Repair service including parts and labor for iPad  149.99         0   
212555  Repair service including parts and labor for iPad  149.99         0   

            type       id  id_order  product_id  product_quantity  unit_price  \
0           8696  1121326    300551           0                 1       54.99   
1           8696  1133226    304067           0                 1       49.99   
2           8696  1134280    304484           0                 1       49.99   
3           8696  1136418    305406           0                 1       49.99   
4           8696  1136825    305590           0                 1       49.99   
...          ...      ...       ...         ...               ...         ...   
212551  1,44E+11  1246243    355016           0                 1      149.99   
212552  1,44E+11  1267651    365406           0                 1      149.99   
212553  1,44E+11  1371194    409162           0                 1      149.99   
212554  1,44E+11  1616673    512821           0                 1      149.98   
212555  1,44E+11  1637554    522044           0                 1      149.98   

                      date  
0      2017-01-02 13:34:30  
1      2017-01-07 09:02:08  
2      2017-01-07 21:17:55  
3      2017-01-09 07:45:12  
4      2017-01-09 11:53:15  
...                    ...  
212551 2017-05-11 18:30:11  
212552 2017-06-13 22:37:38  
212553 2017-10-06 00:32:29  
212554 2018-02-16 12:11:38  
212555 2018-03-06 08:49:24  

[212556 rows x 12 columns]

In [116]:
merge_orderline_products.duplicated().sum()

0

In [117]:
merge_orderline_products.isna().sum()

sku                  0
name                 0
desc                10
price                0
in_stock             0
type                 0
id                   0
id_order             0
product_id           0
product_quantity     0
unit_price           0
date                 0
dtype: int64

In [118]:
ord = merge_orderline_products.merge(orders, left_on = 'id_order', right_on = 'order_id')
ord

sku                                         name  \
0       RAI0007            Silver Rain Design mStand Support   
1       RAI0007            Silver Rain Design mStand Support   
2       RAI0007            Silver Rain Design mStand Support   
3       RAI0007            Silver Rain Design mStand Support   
4       SAT0013  Satechi Support Macbook Aluminum Gray Space   
...         ...                                          ...   
212362  REP0076     repair Full screen iPad (1st generation)   
212363  REP0076     repair Full screen iPad (1st generation)   
212364  REP0076     repair Full screen iPad (1st generation)   
212365  REP0076     repair Full screen iPad (1st generation)   
212366  REP0076     repair Full screen iPad (1st generation)   

                                                     desc   price  in_stock  \
0            Aluminum support compatible with all MacBook   59.99         1   
1            Aluminum support compatible with all MacBook   59.99         1   
2            Aluminum support compatible with all MacBook   59.99         1   
3            Aluminum support compatible with all MacBook   59.99         1   
4       Support Macbook with aluminum finish and foldi...   39.99         1   
...                                                   ...     ...       ...   
212362  Repair service including parts and labor for iPad  149.99         0   
212363  Repair service including parts and labor for iPad  149.99         0   
212364  Repair service including parts and labor for iPad  149.99         0   
212365  Repair service including parts and labor for iPad  149.99         0   
212366  Repair service including parts and labor for iPad  149.99         0   

            type       id  id_order  product_id  product_quantity  unit_price  \
0           8696  1121326    300551           0                 1       54.99   
1           8696  1133226    304067           0                 1       49.99   
2           8696  1134280    304484           0                 1       49.99   
3           8696  1136418    305406           0                 1       49.99   
4           8696  1136416    305406           0                 1       29.99   
...          ...      ...       ...         ...               ...         ...   
212362  1,44E+11  1246243    355016           0                 1      149.99   
212363  1,44E+11  1267651    365406           0                 1      149.99   
212364  1,44E+11  1371194    409162           0                 1      149.99   
212365  1,44E+11  1616673    512821           0                 1      149.98   
212366  1,44E+11  1637554    522044           0                 1      149.98   

                      date  order_id        created_date  total_paid  \
0      2017-01-02 13:34:30    300551 2017-01-02 13:37:16       54.99   
1      2017-01-07 09:02:08    304067 2017-01-07 09:02:08       49.99   
2      2017-01-07 21:17:55    304484 2017-01-07 21:17:55       49.99   
3      2017-01-09 07:45:12    305406 2017-01-09 07:43:42       79.98   
4      2017-01-09 07:43:42    305406 2017-01-09 07:43:42       79.98   
...                    ...       ...                 ...         ...   
212362 2017-05-11 18:30:11    355016 2017-05-11 18:36:09      159.98   
212363 2017-06-13 22:37:38    365406 2017-06-13 22:37:38      219.98   
212364 2017-10-06 00:32:29    409162 2017-10-06 00:32:29      149.99   
212365 2018-02-16 12:11:38    512821 2018-02-16 12:11:38      149.98   
212366 2018-03-06 08:49:24    522044 2018-03-06 08:52:18      159.97   

                  state  
0             Completed  
1       Shopping Basket  
2       Shopping Basket  
3       Shopping Basket  
4       Shopping Basket  
...                 ...  
212362          Pending  
212363  Shopping Basket  
212364  Shopping Basket  
212365  Shopping Basket  
212366          Pending  

[212367 rows x 16 columns]

In [119]:
ord.isna().sum()

sku                  0
name                 0
desc                10
price                0
in_stock             0
type                 0
id                   0
id_order             0
product_id           0
product_quantity     0
unit_price           0
date                 0
order_id             0
created_date         0
total_paid           0
state                0
dtype: int64

In [120]:
ord.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212367 entries, 0 to 212366
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   sku               212367 non-null  object        
 1   name              212367 non-null  object        
 2   desc              212357 non-null  object        
 3   price             212367 non-null  float64       
 4   in_stock          212367 non-null  int64         
 5   type              212367 non-null  object        
 6   id                212367 non-null  int64         
 7   id_order          212367 non-null  int64         
 8   product_id        212367 non-null  int64         
 9   product_quantity  212367 non-null  int64         
 10  unit_price        212367 non-null  float64       
 11  date              212367 non-null  datetime64[ns]
 12  order_id          212367 non-null  int64         
 13  created_date      212367 non-null  datetime64[ns]
 14  tota

In [124]:
orders.groupby("state")["total_paid"].mean()

state
Cancelled          766.738125
Completed          341.381381
Pending            327.208899
Place Order        417.969775
Shopping Basket    729.252943
Name: total_paid, dtype: float64

In [125]:
ord.groupby("state")["total_paid"].mean()

state
Cancelled          350.258639
Completed          224.767206
Pending            205.999355
Place Order        295.727501
Shopping Basket    303.777362
Name: total_paid, dtype: float64